src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,126 @@
+                        int number = Integer.valueOf(num);
+                        if (number <= 0 || number > tasks.getSize()) {
+                            throw new InvalidInputException(str);
+                        }
+                        int index = number - 1; //index for task list
+                        tasks.markDone(index);
+                        Task done = tasks.getTask(index);
+                        ui.printDone(done);
+                    } else if (str.startsWith("unmark ")) {
+                        String num = str.substring(7);
+                        int number = Integer.valueOf(num);
+                        if (number <= 0 || number > tasks.getSize()) {
+                            throw new InvalidInputException(str);
+                        }
+                        int index = number - 1; //index for task list
+                        tasks.markNotDone(index);
+                        Task notDone = tasks.getTask(index);
+                        ui.printNotDone(notDone);
+                    } else if (str.startsWith("delete ")) {
+                        String num = str.substring(7);
+                        int number = Integer.valueOf(num);
+                        if (number <= 0 || number > tasks.getSize()) {
+                            throw new InvalidInputException(str);
+                        }
+                        int index = number - 1;
+                        Task toBeDeleted = tasks.removeTask(index);
+                        ui.printDelete(toBeDeleted, tasks);
+                    } else {
+                        if (str.startsWith("todo")) {
+                            String todo = str.substring(4);
+                            //remove any leading and trailing whitespace characters and
+                            //check whether there is a task after the instruction
+                            if (todo.trim().isEmpty()) {
+                                //this would mean the instruction is incomplete
+                                throw new ToDoCommandUseException(str);
+                            }
+                            String string = str.substring(5);
+                            Task task = new ToDo(string);
+                            tasks.addTask(task);
+                            ui.printAddTask(task, tasks);
+                        } else if (str.startsWith("deadline")) {
+                            if (!str.contains("/by ")) {
+                                throw new DeadlineCommandUseException(str); //needs to check for /by
+                            } else {
+                                String byWhen = "/by ";
+                                int index = str.indexOf(byWhen);
+                                String deadline = str.substring(index + 4); //remove /by from the substring
+                                if (deadline.trim().isEmpty()) { //needs to check whether there is anything after /by
```

Comment:
Is there any considerations as to the deep nesting of logic here? Perhaps consider avoid deep nesting within this method as it makes keeping track of logic difficult! 😊